### ISSUE TITLE
T-SQL (APPLY ISSUES PARSING)

### ISSUE THREAD

**Issue by extremesecrecy** [tobymao/sqlglot#5874]
SQLGlot OUTER APPLY Limitations Bug Report

**Date:** 2025-01-11  
**Reporter:** Gabriel
**Project:** sql BURSTING  
**Affected Systems:** comprehensive_sql_validator.py, UnifiedGeneratorAPI  
**Severity:** High - Impacts validation accuracy and query generation

   Executive Summary

SQLGlot's T-SQL dialect parsing has fundamental limitations when handling OUTER APPLY and CROSS APPLY constructs, causing false positive validation errors and preventing proper query analysis. This affects both the comprehensive SQL validator and the UnifiedGeneratorAPI's ability to parse and generate complex Gabes Database queries.

   Problem Description

    Core Issue
SQLGlot fails to properly understand T-SQL scoping semantics in OUTER APPLY constructs, interpreting table aliases defined within OUTER APPLY blocks as missing JOINs rather than recognizing them as valid scoped references.

    Impact Assessment
- **Validation System:** 30 false positive "missing JOIN" errors across 4 SQL files (98% accuracy instead of 100%)
- **Query Generation:** UnifiedGeneratorAPI cannot properly parse complex queries with OUTER APPLY for template creation
- **Training Data:** Affects STC pipeline's ability to validate generated SQL variations

   Technical Details

    SQLGlot Dialect Configuration
Current implementation uses T-SQL dialect specification:
```python
roots = ast_core.parse_many(content, dialect='tsql', auto=False)
```

    T-SQL OUTER APPLY vs SQLGlot Understanding

     Valid T-SQL Pattern
```sql
FROM 
    HOMEHEALTH_DATA hei
INNER JOIN 
    EXPERIENCE epi ON epi.EXPERIENCE_ID = hei.EXPERIENCE_ID
OUTER APPLY (
    SELECT TOP 1 
        ha3.BILLING_ID
    FROM 
        BILLING3 ha3
    INNER JOIN 
        BILLING ha ON ha.BILLING_ID = ha3.BILLING_ID
    WHERE 
        ha3.EXPERIENCE_ID = hei.EXPERIENCE_ID  -- Valid reference to outer scope
) acc
```

     SQLGlot's Misinterpretation
- **Expected:** SQLGlot should recognize `hei` as available in OUTER APPLY scope
- **Actual:** SQLGlot flags `hei` as undefined, reporting "missing JOIN"
- **Result:** False positive validation error

   Specific Examples from Affected Files

    Example 1: 30_HomeHealthAdminData.sql (Lines 44-55)
**OUTER APPLY construct flagged as missing JOIN:**
```sql
OUTER APPLY (
    SELECT TOP 1 ha3.BILLING_ID
    FROM BILLING3 ha3
    INNER JOIN BILLING ha ON ha.BILLING_ID = ha3.BILLING_ID
    WHERE ha3.EXPERIENCE_ID = hei.EXPERIENCE_ID
) acc
```
**False Positive:** Validator reports `ha` and `ha3` as missing JOINs
**Reality:** These are properly scoped within the OUTER APPLY block

    Example 2: 219_CarePlanTask.sql (Lines 52-59)
**OUTER APPLY with MIN aggregation:**
```sql
OUTER APPLY (
    SELECT MIN(addd.DUE_DATE_DTTM) AS DUE_DATE
    FROM BILLING_DUE addd
    WHERE addd.ACTIVITY_ID = tpa.ACTIVITY_ID
) due
```
**False Positive:** Validator reports `addd` as missing JOIN
**Reality:** `addd` is the correct alias for BILLING_DUE within OUTER APPLY scope

    Example 3: 97_HomeHealthVerbalOrders.sql (Multiple OUTER APPLY blocks)
**Pattern with correlated subquery:**
```sql
OUTER APPLY (
    SELECT TOP 1 
        hve.DATE_OF_EVENT as VERBAL_ORDER_END_DATE
    FROM EVENT_DATES hve
    WHERE hve.DATE_EVENT_C = 3 
        AND hve.EXPERIENCE_ID = hei.EXPERIENCE_ID
) verbal_end
```
**False Positive:** Validator reports `hve` as missing JOIN
**Reality:** Proper correlated subquery pattern in T-SQL

   Attempted Solutions and Their Limitations

    Solution 1: Enhanced Scope Analysis
**Approach:** Modified validator to track OUTER APPLY scope and mark aliases as valid
```python
def analyze_outer_apply_scope(self, node):
    """Enhanced OUTER APPLY scope analysis"""
    if hasattr(node, 'this') and isinstance(node.this, exp.Select):
          Track aliases within OUTER APPLY scope
        for table in node.this.find_all(exp.Table):
            if table.alias:
                self.outer_apply_aliases.add(table.alias.upper())
```
**Limitation:** SQLGlot's AST structure doesn't preserve proper scoping semantics for T-SQL constructs

    Solution 2: T-SQL Dialect Specification
**Approach:** Explicitly set T-SQL dialect and disable auto-detection
```python
roots = ast_core.parse_many(content, dialect='tsql', auto=False)
```
**Limitation:** SQLGlot's T-SQL parser still treats OUTER APPLY as standard LEFT JOIN LATERAL, losing T-SQL-specific scoping rules

    Solution 3: Manual Exception Handling
**Approach:** Create allowlist of known OUTER APPLY patterns
```python
def is_outer_apply_false_positive(self, alias, line_content):
    """Check if missing JOIN is actually valid OUTER APPLY pattern"""
    return ('OUTER APPLY' in line_content.upper() and 
            any(pattern in line_content.upper() for pattern in self.outer_apply_patterns))
```
**Limitation:** Requires manual maintenance and doesn't solve the root parsing issue

   Root Cause Analysis

    SQLGlot Architecture Limitation
SQLGlot converts T-SQL constructs to a standardized AST, but:
1. OUTER APPLY gets converted to LEFT JOIN LATERAL equivalents
2. T-SQL scoping rules are not preserved in the conversion
3. The parser loses context about which aliases are available in which scope levels

    T-SQL vs Standard SQL Semantics
- **T-SQL OUTER APPLY:** Creates a new scope where outer query aliases are accessible
- **Standard SQL LATERAL:** Different scoping rules that SQLGlot applies uniformly
- **Result:** Semantic mismatch causing validation errors

   Impact on Systems

    Comprehensive SQL Validator
- **Current State:** 164/168 files pass validation (97.6% accuracy)
- **False Positives:** 30 missing JOIN errors across 4 files
- **Business Impact:** Cannot achieve 100% validation confidence for STC pipeline

    UnifiedGeneratorAPI
- **Query Parsing:** Cannot properly parse complex Gabes Database queries with OUTER APPLY
- **Template Generation:** Fails to create accurate query templates from existing SQL
- **Variation Generation:** Cannot generate valid SQL variations that preserve OUTER APPLY semantics

   Recommended Solutions

    Short-term (Immediate)
1. **Validation Bypass:** Create specific exception handling for known OUTER APPLY patterns
2. **Manual Verification:** Document the 4 files with false positives as manually verified
3. **Pipeline Continuation:** Proceed with STC data generation using current 97.6% validation rate

    Medium-term (Next Sprint)
1. **Alternative Parser:** Evaluate Microsoft's T-SQL parser or sqlparse library
2. **Hybrid Approach:** Use SQLGlot for standard constructs, specialized parser for T-SQL-specific patterns
3. **Custom AST Enhancement:** Extend SQLGlot's T-SQL dialect with proper OUTER APPLY handling

    Long-term (Future Roadmap)
1. **Contribute to SQLGlot:** Submit PR to improve T-SQL OUTER APPLY handling
2. **Custom SQL Engine:** Develop Gabes Database-specific SQL parser for medical domain queries
3. **API Abstraction:** Create abstraction layer that handles different SQL dialects transparently

   Testing Strategy

    Validation Tests
```sql
-- Test Case 1: Simple OUTER APPLY
FROM TableA a
OUTER APPLY (SELECT x FROM TableB b WHERE b.id = a.id) sub

-- Test Case 2: Nested OUTER APPLY
FROM TableA a
OUTER APPLY (
    SELECT x FROM TableB b 
    OUTER APPLY (SELECT y FROM TableC c WHERE c.b_id = b.id) sub2
    WHERE b.a_id = a.id
) sub1

-- Test Case 3: Multiple OUTER APPLY
FROM TableA a
OUTER APPLY (SELECT x FROM TableB b WHERE b.id = a.id) sub1
OUTER APPLY (SELECT y FROM TableC c WHERE c.id = a.id) sub2
```

   Files Requiring Attention

    Primary Files with False Positives
1. **30_HomeHealthAdminData.sql** - 17 missing JOIN errors (all OUTER APPLY related)
2. **219_CarePlanTask.sql** - 4 missing JOIN errors (OUTER APPLY with aggregations)
3. **97_HomeHealthVerbalOrders.sql** - 8 missing JOIN errors (complex OUTER APPLY patterns)
4. **98_CashDrawerDeposits.sql** - 1 missing JOIN error (simple OUTER APPLY)

    System Files Affected
1. **scripts/comprehensive_sql_validator.py** - Validation logic needs OUTER APPLY exception handling
2. **UnifiedGeneratorAPI** - Query parsing and template generation affected
3. **STC Pipeline** - Training data generation validation step impacted

   Conclusion

SQLGlot's T-SQL dialect limitations represent a significant technical debt that affects multiple systems in the STC pipeline. While workarounds exist for immediate needs, a long-term solution requiring either alternative parsing libraries or contributions back to the SQLGlot project is necessary for achieving 100% validation accuracy and full query generation capabilities.

The current 97.6% validation rate is acceptable for proceeding with STC data generation, but the 30 false positive errors must be documented and manually verified to ensure training data quality.

---
**Next Steps:**
1. Implement short-term validation bypass for the 4 affected files
2. Document manual verification of OUTER APPLY constructs in these files  
3. Continue STC pipeline development with current validation constraints
4. Evaluate alternative T-SQL parsers for future implementation

**Comment by georgesittas** [comment]
Hey @extremesecrecy, appreciate the additional details here, but this issue is incomprehensible. Can you please try to simplify it a bit? Issues should ideally look like:

```python
import sqlglot
sqlglot.parse_one(<a simplified version of your query that is as small as possible>, read="tsql")  # this fails
```

**Comment by extremesecrecy** [comment]
FROM 
    HOMEHEALTH_DATA hei
INNER JOIN 
    EXPERIENCE epi ON epi.EXPERIENCE_ID = hei.EXPERIENCE_ID
OUTER APPLY (
    SELECT TOP 1 
        ha3.BILLING_ID
    FROM 
        BILLING3 ha3
    INNER JOIN 
        BILLING ha ON ha.BILLING_ID = ha3.BILLING_ID
    WHERE 
        ha3.EXPERIENCE_ID = hei.EXPERIENCE_ID  -- Valid reference to outer scope
) acc

 SQLGlot's Misinterpretation
Expected: SQLGlot should recognize hei as available in OUTER APPLY scope

Actual: SQLGlot flags hei as undefined, reporting "missing JOIN"

The issue is that SQLGlot takes the outer apply and converts it into a left lateral join and when you compare the joins, it does not allow to make sure that we can validate that the join is correct because the interpreted AST is different than the expected TSQL. I.E. Takes an apple, transform it to an orange, then we want to compare if the apple is the right kind of an apple, but now, it is always not the right apple because it is an orange. I think my example might not be great because both apple and orange can do the same "function", I guess juice, but when the job is to compare and see if the joins are valid, this really makes it hard to do it natively because of the transformation.

**Comment by geooo109** [comment]
The query is correctly transpiled. SQLGlot needs a generic way to handle transpilation. I believe you should work with it as it is. If you find any wrong transpilations let us know.

**Comment by extremesecrecy** [comment]
@geooo109 , are you saying that once I AST my query, to compare, I have to transpilate it to T-sql and then compare?
